### PR TITLE
fix(node-resolve): mark module as external if resolved module is external

### LINKED ISSUE #609: External modules are resolved / re-written instead of being ignored
- Rollup Plugin Name: @rollup/plugin-node-resolve
- Rollup Plugin Version: 9.0.0
- Rollup Version: 2.30.0
- Operating System (or Browser): macOS 10.15.7
- Node Version: 10.22.0
- Link to reproduction: https://repl.it/join/xdisrknw-matthewdean2

### Expected Behavior
- External modules should not be resolved / re-written, as they are, by definition, external

### Actual Behavior
- External modules are re-written with a `/node_modules/` path. This prevents the Rollup-bundled library from being imported as a module into another repo, as dependencies will not be local to the _original_ repo, they will be installed as dependencies in the importing repo.

For example this line:
`import axios from 'axios'`
should remain:
`import axios from 'axios'`

Instead, it turns into:
`import axios from './node_modules/axios/index.js';`

### Additional Information

Any temporary workarounds to this behavior would be welcome.

### PR DESCRIPTION
**Pull request by tjenkinson**
## Rollup Plugin Name: `node-resolve`

This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [x] yes (_breaking changes will not be merged unless absolutely necessary_)
- [ ] no

List any relevant issue numbers: fixes #609

### Description

**Recommend telling github to ignore the whitespace changes for review.** A lot of the code just moved location

If the ID given to this plugin resolves to something which itself resolves to an external, then we should mark the ID the plugin got as external. As a user I think I'd expect `node-resolve` to either resolve to a file, or leave the import alone. Not sure resolving to something that's then an external would be a normal use case?

There's a special case where `this.resolve()` might result in another plugin (i.e. commonjs) also doing a `this.resolve()` causing an infinte loop. I added some logic to handle this case. Not sure if there's a better way?

### CODE REVIEW

File: packages/node-resolve/src/index.js [hunk 1]
```diff
@@ -71,201 +72,220 @@ export function nodeResolve(opts = {}) {
...
-        if (importeeIsBuiltin && preferBuiltins) {
-          if (!isPreferBuiltinsSet && resolvedWithoutBuiltins && resolved !== importee) {
-            this.warn(
-              `preferring built-in module '${importee}' over local alternative at '${resolvedWithoutBuiltins.location}', pass 'preferBuiltins: false' to disable this behavior or 'preferBuiltins: true' to disable this warning`
-            );
+      const resolved = await doResolveId(this, importee, importer, opts);
+      if (resolved && !inNestedResolve) {
+        // if a plugin invoked by `this.resolve` also does `this.resolve`
+        // we need to break the loop
+        inNestedResolve = true;
```
Comment (lukastaegert) [inline packages/node-resolve/src/index.js:277]: I do not think having a global flag here is a good idea. As resolving is inherently asynchronous, concurrent resolves can mess up the value of this flag. I fear this problem needs to be solved in Rollup core by correctly tracking uses of the `skipSelf` flag: If a plugin uses `this.resolve` with the `skipSelf` flag, then Rollup needs to remember this when calling `resolveId` of other plugins so that when another plugin subsequently calls `this.resolve` for THE SAME ID (that is important), the original plugin is still skipped (and this should be remembered transitively). I believe this is the issue you are trying to solve?
I thought about this myself some time ago but did not implement anything yet, maybe I should.

**Review by lukastaegert — COMMENTED** [review]
Looks mostly good but I am not happy about the loop prevention mechanism.

File: packages/node-resolve/src/index.js (hunk 1 above)
Comment (tjenkinson) [reply packages/node-resolve/src/index.js:277]: Ah yep. I was thinking it would only be called once at a time on a single instance, but wasn't thinking about multiple ids in parallel.

>I believe this is the issue you are trying to solve?

I think that would work for this case, although the difference with what you proposed is that the `node-resolve` plugin would be skipped completely when `this.resolve` invoked from commonjs meaning later plugins might get invoked that wouldn't usually, vs now where it would still stop when `node-resolve` returns something (although what it returns now is not really the correct thing, given it could map to an external).

Not sure where the best place to break the chain is, and also when breaking the chain if there is a sensible value that could be returned or if we should actually signal an infinite loop back to the thing that called `this.resolve`.

Confusing problem to think about :)

Comment (lukastaegert) [reply packages/node-resolve/src/index.js:277]: True. But to my understanding, this change is not essential to the functionality you are implementing, right? So you could remove it from here without any regressions and one could always add something later, but it here or in Rollup core.

Comment (tjenkinson) [reply packages/node-resolve/src/index.js:277]: Not quite understanding. The change here to do the nested `this.resolve` is needed (unless there is another solution I'm missing) to fix #609 

I don't mind if the fix is entirely here (with a tweak to keep a `inNestedResolve` flag against the id) or partly here (to do the `this.resolve`) but with a core change to prevent the loops.

Also this isn't a blocker for me btw. I just saw the issue and decided to attempt to fix it, so not sure it's a particularly urgent problem to solve.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js:277]: Ah sorry, should have read the entire thread. The problem is that this is an issue in core that affects all plugins. But to fix it just for node-resolve, you could maintain a data structure that contains the source/importer of what we are just resolving to remember to skip exactly this call when resolving. From Rollup side, each combination will be resolved exactly once. But you could clean up the data structure once a resolution is done to account for plugins resolving the same modules at a later stage.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js:277]: If we later fix this in core, then this fix should not pose a problem as it will just become inactive.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js:277]: A solution from Rollup core would be similar: When skipSelf is used, all nested resolve calls would skip the plugin if the same combination of source and importer is used, but call the plugin for different combinations

Comment (tjenkinson) [reply packages/node-resolve/src/index.js]: :+1: I updated the PR to check the `importee` and `importer`.

I'm not sure this is something the core could handle. For this case I _think_ we need to return what the result would be from the nested `resolveId` call, not `null`, because our `this.resolve` is answering the question "If we were to return the result from the node resolve algo, would something else make it external?". If we return `null`, we're not satisfying "If we were to return the result from the node resolve algo".

If we did return `null`

>package-entry-points › uses "require" condition when a module is referenced with require

fails with `CIRCULAR_DEPENDENCY` (in addition to the new test)

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: > I'm not sure this is something the core could handle. For this case I think we need to return what the result would be from the nested resolveId call, not null, because our this.resolve is answering the question "If we were to return the result from the node resolve algo, would something else make it external?". If we return null, we're not satisfying "If we were to return the result from the node resolve algo".

While you are correct that in general that this might cause edge case issues, it would still be correct here as we are calling this.resolve with the resolved id. So the algorithm I have in mind for Rollup core would still pass it to node-resolve once, and even if it would not, node-resolve would not change a resolved id anyway.

I think I will still add this to core eventually because it fixes a lot of fundamental issues with the skipSelf approach (and would only affect secondary `this.resolve` calls that I would assume are usually unlikely to change the actual `resolveId` result, even though this is mostly a heuristic).

**Review by lukastaegert — APPROVED** [review]
Looks good from my side

File: packages/node-resolve/src/index.js (hunk 1 above)
Comment (tjenkinson) [reply packages/node-resolve/src/index.js]: Maybe I'm not understanding what the algorithm would be

If we have
```
commonjs this.resolve('a') -> node-resolve this.resolve('b') -> commonjs this.resolve('b') -> node-resolve this.resolve('b')
```

(where node-resolve resolves 'a' to 'b')

and that last call returns `null` not `'b'`, that's a problem because commonjs thinks nothing can resolve 'a'.

With your algorithm would that last `this.resolve(b)` just return `'b'` immediately?

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: > that's a problem because commonjs thinks nothing can resolve 'a'

Actually there is always the core resolver. Which would just resolve an absolute URL to itself: https://github.com/rollup/rollup/blob/9bccc4fee3ee19a6f5d252cc4ccbd174b46f6a5d/src/utils/resolveId.ts#L24-L24

There would only be an issue in the browser, but none of those plugins works with the browser build anyway.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: So in your example, "b" would be an absolute path and the last resolve would again resolve it to "b"

Comment (tjenkinson) [reply packages/node-resolve/src/index.js]: Ah got it. There's still the case where if a later plugin does actually resolve to something else that could cause an issue though. Also updating the code to

```ts
    const resolved = await doResolveId(this, importee, importer, opts);
      if (resolved) {
        const resolveContext = { importee: resolved.id, importer };
        // Would the result from the node resoulution algorithm get marked as external by another plugin (or config)?
        // If that's the case we want to mark the import as external to keep the original import intact.
        if (!nestedResolveTracker.inNestedResolve(resolveContext)) {
          const isExternal = await nestedResolveTracker.track(resolveContext, async () => {
            const resolvedResolved = await this.resolve(resolved.id, importer, { skipSelf: true });
            return !!(resolvedResolved && resolvedResolved.external);
          });
          if (isExternal) {
            return false;
          }
        } else {
          // added this
          return null;
        }
```
which I think is going in the direction you're thinking for the core does cause a test failure that has a mix of commonjs and node-resolve.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: No, it is slightly different. What I am suggesting is to never call `resolveId` with the same arguments twice for any plugin in nested `resolve` calls. In your code, this would be equivalent to

```js
if (resolved) {
  // Would the result from the node resoulution algorithm get marked as external by another plugin (or config)?
  // If that's the case we want to mark the import as external to keep the original import intact.
  if (!nestedResolveTracker.inNestedResolve({ importee, importer })) {
    const isExternal = await nestedResolveTracker.track(
      { importee: resolved.id, importer },
      async () => {
        const resolvedResolved = await this.resolve(resolved.id, importer, {
          skipSelf: true
        });
        return !!(resolvedResolved && resolvedResolved.external);
      }
    );
    if (isExternal) {
      return false;
    }
  } else {
    return null;
  }
}
```

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: So `resolveId` would indeed be called twice, first with the unresolved source and then with the resolved source. Only the third call is skipped when it is called with the resolved source again. This fixes the test in question and only breaks the test that specifically looks for this.
I know it can still cause unexpected situations but I do not think them very likely. But the alternative is that every plugin that uses `this.resolve` needs an elaborate mechanism to guard against this. If there is demand, we could also add another flag to `this.resolve` to allow plugins to specify that they can handle this on their own and do not want the default circularity prevention.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: I created a preliminary patch in rollup/rollup#4000 to play around with.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: I have extended the PR in rollup/rollup#4000 to include all the necessary tests and also extended documentation. The idea here would be that if you do not want this behaviour, your "opt-out" is to not use `skipSelf` but entirely implement your own mechanism.
I would be willing to release it if there are no hard concerns against this in a minor release.
Note however that if we want to use this in the node-resolve plugin instead of the custom logic (which I think is a good idea as it reduces complexity), then this would mean changing the peer dependency version of Rollup which would be a breaking change.

Comment (tjenkinson) [reply packages/node-resolve/src/index.js]: I'm fine with removing that extra test and simplifying the plugin to use the core change. The edge case is so many layers deep to think about it does seem unlikely, and there's still the way of optiing out if needed in the future like you mentioned in the docs.

It's bugging me though I still can't actually understand why the extra loop round is fixing the test with commonjs 🤔

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: Maybe the question is more why it was broken in the first place without the second round. The reason is that the CommonJS plugin was doing this:

```
this.resolve(importee, importer, {
      skipSelf: true,
      custom: { 'node-resolve': { isRequire: isProxyModule || isRequiredModule } }
    })
```

passing special options to the node-resolve plugin to change resolution, and this is the call that was skipped. Admittedly we could also try to hash the custom options object in the loop prevention, but this is dangerous as we cannot go for object identity as it is likely recreated for each call, and there is no requirement to have it serialisable via JSON stringify.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: We could do a JSON.stringify in a try-catch as a "best effort" approach.

Comment (lukastaegert) [reply packages/node-resolve/src/index.js]: But now I have written this nice description and I am feeling the actual logic is getting more and more complex. Maybe we start with what we have now and see if this causes issues anywhere.

Comment (tjenkinson) [reply packages/node-resolve/src/index.js]: Ahhh thanks that was the missing piece. Agree ideally the `custom` should also be part of the key, but waiting until that actually causes a problem (if it does) seems reasonable. Will update this PR

**Review by lukastaegert — APPROVED** [review]

**Comment by lukastaegert** [comment]
Looks good, thanks for putting in the work!

**Review by shellscape — APPROVED** [review]

**Comment by lazka** [comment]
The following now fails with `TypeError: this.resolve is not a function`

```js
import resolve from '@rollup/plugin-node-resolve';
await resolve().resolveId("rollup");
```

Any ideas what I'm doing wrong here? :/

**Comment by lazka** [comment]
I see. I was using the plugin API to figure out the path of packages so I can copy assets from them. If that's not supported feel free to ignore.

**Comment by lazka** [comment]
> Ok. Could you use `require.resolve` in node or maybe the `resolve` package directly?

When I last tried I couldn't get it to work for non-js packages, but I now see that `require.resolve(packageName + '/package.json')` also works. I'll give that a try. Thank you!

**Comment by Waidmann** [comment]
Lgtm
